Add function types and C++ reference types to Pure Python mode

dgrunwald-qt opened the pull request:
Adds new syntax for function types (#4279) and C++ reference types:
* `cython.pointer[cython.function_type([cython.int], cython.double)]` -> `double (*)(int) except *`
* `cython.reference[T]` -> `T&`
* `cython.rvalue_reference[T]` -> `T&&`

dgrunwald-qt committed:
Add C++ reference types to Pure Python mode.

dgrunwald-qt committed:
#4279: Add function types to Pure Python mode.

scoder reviewed (commented):
Cool, thanks for writing this up.

Let's discuss naming issues in the ticket.

dgrunwald-qt commented:
Well, arguably C++ does not have syntax for forwarding references, they're an emergent feature from the combination of rvalue references, type deduction, and reference collapsing.
I think I still prefer `rvalue_reference` even if they are not fully implemented yet. But I could also omit them for now. My own need is only for normal reference types, I just saw that `PyrexTypes.CppRvalueReferenceType` exists and so added `rvalue_reference` for completeness.

dgrunwald-qt committed:
Changed `function_type` API based on feedback; applied reviewer comments.

scoder commented on Cython/Compiler/ExprNodes.py:
```diff
@@ -6158,6 +6164,77 @@ def set_py_result_type(self, function, func_type=None):
+                        elif isinstance(v, bool):
+                            exc_check = '+' if v else False
+                        else:
+                            error(self.pos, f"Value for {k} must be '*', string or boolean, was {v!r}")
```
```suggestion
                            error(kv_pair.value.pos, f"Value for {k} must be '*', string or boolean, was {v!r}")
```

scoder commented on Cython/Compiler/ExprNodes.py:
```diff
@@ -6158,6 +6164,77 @@ def set_py_result_type(self, function, func_type=None):
+
+                    else:
+                        error(self.pos, f"Unknown kwarg in function_type: {k}")
+            if noexcept and exc_clause:
+                error(self.pos, "Cannot combine noexcept=True with another exception clause.")
```
We could provide a better error position here if we detected this case inside of the loop rather than afterwards.

scoder reviewed (commented):
Mostly ok, but some semantics still need adjustment.

Maybe it would help to have a test that directly compares the Cython syntax declaration with the corresponding Python declaration? Maybe even making sure that suitable `cfunc` functions can correctly be assigned to both pointers?

dgrunwald-qt committed:
Rename except_plus->except_cpp

dgrunwald-qt committed:
Add comparison test.

dgrunwald-qt committed:
Fix implicit conversion of except_val to function return type

dgrunwald-qt committed:
Improve error positions

scoder committed:
Merge branch 'master' into pure-python-reference-types

scoder committed:
Rename pure_pyx_cpp.pyx to pure_funcptr_cpp.pyx

dgrunwald-qt committed:
Merge remote-tracking branch 'upstream/master' into pure-python-reference-types

matusvalo commented:
I think this PR should introduce also a documentation of new constructs...

matusvalo committed:
Merge branch 'master' into pure-python-reference-types

da-woods commented:
To me this looks largely good. I agree with Matus that a little bit of documentation in https://cython.readthedocs.io/en/latest/src/tutorial/pure.html#c-types would be nice.

I wonder if it'd make sense to use `...` at the end of the list of arguments to annotation var-args (as opposed to a keyword argument). I think that'd be fairly easy to detect (it's just `EllipsisNode`) and it'd be consistent with `typing.Callable` (as far as it's meaningful to be).

scoder reviewed (commented):
Yes, documentation is definitely missing.

This is also lacking error tests. There is a bunch of new errors, most of which can be tested.

And a running test for the new features would help. The added tests are only printing types, but are not executing anything that would exercise the code that we generate for the declarations.

scoder committed:
Merge branch 'master' into pure-python-reference-types

dgrunwald-qt committed:
Add documentation for function_type and reference.

dgrunwald-qt committed:
Update Shadow.py to allow type-checking function_type usages.

With this commit, pyright accepts

    FnPtr = cython.typedef[cython.pointer[cython.function_type([cython.int], cython.void, nogil=True)]]

dgrunwald-qt committed:
Replace `has_varargs=True` with `...` as last parameter.

scoder committed:
Merge branch 'master' into pure-python-reference-types

da-woods commented:
Would `cython.exceptval[Literal[-1]]` get through type checkers? (If so, then my opinion would be to support it for people who care about that, and accept `cython.exceptval[-1]` for people who prefer brevity)
